### PR TITLE
#1 - Allow multiple idp's with the same metadatalink

### LINKED ISSUE #1: Allow multiple idp's with the same metadatalink
KS laring have requested us to make it so you can have multiple identity providers with the same metadatalink in the auth2_saml plugin.

It's catalyst's plugin and we want to try and merge it into their version. Until then ks will use our fork

**Remember to register time spent!**

### PR DESCRIPTION
**Pull request by Anders-Oerbaek-Pedersen**
Fix for #1 

Notable changes:

* Changed use of internal array key by from non unique database field `entityid` to unique database field `id`

### CODE REVIEW

File: classes/auth.php [hunk 1]
```diff
@@ -123,12 +123,13 @@ public function __construct() {
         foreach ($idpentities as $idpentity) {
             // Set name.
             $idpentity->name = empty($idpentity->displayname) ? $idpentity->defaultname : $idpentity->displayname;
-            $idpentity->md5entityid = md5($idpentity->entityid);
+            $idpentity->md5id = md5($idpentity->entityid);
+            $idpentity->md5id = md5($idpentity->id);
```
Comment (frederikmillingpytlick) [inline classes/auth.php]: Overwriting immediately after?

**Review by frederikmillingpytlick — CHANGES_REQUESTED** [review]

File: classes/auth.php (hunk 1 above)
Comment (Anders-Oerbaek-Pedersen) [reply classes/auth.php]: Forgot to cleanup original line, ended up changing that one also, when renaming object key.

I have removed the superfluous line.

**Review by frederikmillingpytlick — APPROVED** [review]

**Comment by Logiar** [comment]
When saving on /admin/settings.php?section=authsettingsaml2 it does not map them to the old rows, causing new rows to be inserted and old ones to get wiped. This causes a reset to default for settings stored in auth_saml2_idps table. [classes/admin/setting_idpmetadata.php:158](https://github.com/praxisdigital/moodle-auth_saml2/blob/ccf0f4391863837202182e1fb37a87d3af38c99f/classes/admin/setting_idpmetadata.php#L158) is inconsistent with [classes/admin/setting_idpmetadata.php:93](https://github.com/praxisdigital/moodle-auth_saml2/blob/ccf0f4391863837202182e1fb37a87d3af38c99f/classes/admin/setting_idpmetadata.php#L93) and database id can't be used for mapping here.

**Comment by Anders-Oerbaek-Pedersen** [comment]
Mapping should be fixed now.
If the override name is changed, mapping will not match with the previous, causing the old row to be removed and a new row to be inserted.
This is caused by having nothing unique to map to, when allowing two identical IdP metadatalinks and IdP entries being declared as one text input.

**Review by sampraxis — APPROVED** [review]

**Comment by Anders-Oerbaek-Pedersen** [comment]
Tested via show and tell with: JRM

**Pull request merged**